providers/azure.go

```diff
@@ -230,53 +323,32 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 		err = otherMailsErr
 	}
 
+	if err != nil || email == "" {
+		email, err = json.Get("userPrincipalName").String()
+		if err != nil {
+			logger.Errorf("unable to find userPrincipalName: %s", err)
+			return "", err
+		}
+	}
+
 	return email, err
 }
 
-// GetEmailAddress returns the Account email address
-func (p *AzureProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	var email string
-	var err error
-
-	if s.AccessToken == "" {
+func getEmailFromProfileAPI(ctx context.Context, accessToken, profileURL string) (string, error) {
```

Comment:
I think you should use `(p *AzureProvider)` as the method receiver on this so you don't have to explicitly extract and pass the `profileURL`.

You can see nearly identical functionality as this in the OIDCProvider:

```
// enrichFromProfileURL enriches a session's Email & Groups via the JSON response of
// an OIDC profile URL
func (p *OIDCProvider) enrichFromProfileURL(ctx context.Context, s *sessions.SessionState) error {
```